Написана лабораторная №1

eggorrus opened the pull request.

eggorrus committed:
Написана лабораторная №1

BoychikFromAsia commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,68 @@
+
+template <typename T>
+
+class Stack
```
xml документация к типу и всему публичному контракту

eggorrus replied:
сделано

BoychikFromAsia commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,68 @@
+		return size;
+	}
+
+	void clear()
```
Это не очевидно для других разработчиков, что чтобы высвободить память стека надо вызвать clear, во избежание утечек памяти, лучше создать деструктор

eggorrus replied:
деструктор создан.

BoychikFromAsia commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,29 @@
+	Stack<int> myStack;
+	cout << "Введите числа" << endl;
+	cin >> toStackValue;
+	while (toStackValue != 0)
```
0 - окончание последовательности, его не следует помещать в стек

eggorrus replied:
Он и так не будет введен в стек. Если его ввести первым, то условие цикла while сразу же не будет соблюдено. Если же каким-либо номером в последовательности, то после ввода 0 последует следующая итерация цикла while, где будет осуществлена проверка, является ли число 0, и если является, то все, что относится к while, будет проигнорировано

eggorrus committed:
Исправил обработку исключений, обратил внимание на замечания ревьюверов

1. Обработка ошибок реализована через оператор throw;
2. Добавлена xml-документация;
3. Реализован деструктор класса Stack;
4. Значения указателя top и переменной size перенесены в зону public

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+
+using namespace std;
+
+template <typename T>
+
+/// <summary>
+/// ���� ��������� �������� ������ "Stack", ������� �������� �������� �������� ������ ���� "LIFO"
+/// Description of class "Stack" which is data storage "LIFO" system is below.
+/// </summary>
+class Stack
```
Лучше перенести template <typename T> под xml документацию, перед Stack

eggorrus replied:
сделано

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+/// </summary>
+class Stack
+{
+private:
+	struct Node
+	{
+		T data;
+		Node* next;
+	};
+
```
Нужно добавить конструктор к структуре Node

eggorrus replied:
добавлен

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+	/// ���������� � ��������� ����� ��������� �� ������� �����, �� ��������� ���������� nullptr, � ����� ������ ������������� ����������, ���������� �� ������ �����.
+	/// In the public part, we use a pointer to the top of the stack, which is nullptr by default, and also enter an integer variable responsible for the stack size.
+	/// </summary>
+	Node* top = nullptr;
+	int size = 0;
```
Нужно объявлять переменные в приватной секции, а инициализировать в публичной секции, в конструкторе класса Stack

eggorrus replied:
сделано

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+	/// ���������� � ��������� ����� ��������� �� ������� �����, �� ��������� ���������� nullptr, � ����� ������ ������������� ����������, ���������� �� ������ �����.
+	/// In the public part, we use a pointer to the top of the stack, which is nullptr by default, and also enter an integer variable responsible for the stack size.
+	/// </summary>
+	Node* top = nullptr;
+	int size = 0;
```
Лучше выбрать другое название для переменной, например counter

eggorrus replied:
сделано

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+	{
+		clear();
+	}
+	/// <summary>
+	/// ����� ������������ ��������� ������ ��� ������ �� ������:
+	/// void push(T)* - ��������� ������� �� ������� �����
+	/// T pop() * -��������� ������� � ������� ����� � ���������� ��� ��������
+	/// T peek() * -���������� ������� ����������� �� ������� �����
+	/// int count() * -���������� ���������� ��������� � ���������
+	/// void clear() * -������� ��� �������� �� ���������
+	/// The following methods are presented for working with the stack:
+	/// void push(T)* - adds an element to the top of the stack
+	/// T pop() * -retrieves an element from the top of the stack and returns its value
+	/// T peek() * -returns the element at the top of the stack
+	/// int count() * -returns the number of items in the collection
+	/// void clear() * -removes all items from the collection
+	/// </summary>
+	/// <param name="value"></param>
```
Лучше написать документацию к каждому методу отдельно

eggorrus replied:
написал

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+		}
+		else
+		{
+			throw runtime_error("Stack is empty");
```
В данном случае лучше использовать не runtime_error, а out_of_range

eggorrus replied:
сделано

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,106 @@
+		}
+		else
+		{
+			throw runtime_error("Stack is empty");
```
В данном случае лучше использовать не runtime_error, а out_of_range

eggorrus replied:
сделано

GMtimur commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,29 @@
-//   5. Последовательно выберите пункты меню "Проект" > "Добавить новый элемент", чтобы создать файлы кода, или "Проект" > "Добавить существующий элемент", чтобы добавить в проект существующие файлы кода.
-//   6. Чтобы снова открыть этот проект позже, выберите пункты меню "Файл" > "Открыть" > "Проект" и выберите SLN-файл.
+	setlocale(LC_ALL, "");
+	int toStackValue;
```
Лучше изменить название переменной, например inputValue

eggorrus replied:
имя изменено

eggorrus committed:
Исправлены замечания ревьюверов

1. template перенесено под xml документацию
2. Добавлен конструктор к структуре Node
3. Добавлен конструктор к структуре Stack
4. Переменная size переименована в counter
5. Документация прописана к каждому методу отдельно
6. runtime_error заменено на out_of_range
7. toStackValue переименована в inputValue

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+		T data;
+		Node* next;
+
+		Node(const T& value) : data(value), next(nullptr) {} //������ ����������� ��������� Node - created constructor of struct Node
```
Комментарии должны быть на новой строке, по стайл-гайду

eggorrus replied:
исправил

GMtimur commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include <stdexcept>
+#include <string>
```
Для чего была подключена библиотека String?

eggorrus replied:
убрал подключение

eggorrus committed:
Исправил ошибки

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+using namespace std;
+
+/// <summary>
+/// ���� ��������� �������� ������ "Stack", ������� �������� �������� �������� ������ ���� "LIFO"
```
Битая кодировка

eggorrus replied:
исправлено

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+/// </summary>
+
+template <typename T>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+/// ���� ��������� �������� ������ "Stack", ������� �������� �������� �������� ������ ���� "LIFO"
+/// Description of class "Stack" which is data storage "LIFO" system is below.
+/// </summary>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+		Node* next;
+
+		Node(const T& value) : data(value), next(nullptr) {} 
+		//������ ����������� ��������� Node - created constructor of struct Node
```
Комментарии обычно  пишут над тем, о чем идет  речь

eggorrus replied:
Исправлено

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	{
+		clear();
+	}
+	/// <summary>
```
```suggestion
	
	/// <summary>
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	/// The following is the pop method. It extracts and returns a value from the top of the stack.
+	/// </summary>
+	/// <returns></returns>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	/// Method push is below. It adds an element on top of the Stack.
+	/// </summary>
+	/// <param name="value"></param>
+	void push(const T& value)
```
Почему используется ссылка?

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	/// The peek method returns the value of the element located at the top
+	/// </summary>
+	/// <returns></returns>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	/// The count method returns the number of stack elements
+	/// </summary>
+	/// <returns></returns>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,124 @@
+	/// ����� clear ������� ���� � ����������� � �����������.
+	/// The clear method clears the stack and is applied in the destructor.
+	/// </summary>
+
```
```suggestion
```

ichepkov2401 commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,29 @@
+		myStack.push(inputValue);
+		cin >> inputValue;
+	}
+	int currentValue;
```
Слишком большая область видимости

eggorrus replied:
исправлено

ichepkov2401 commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,29 @@
+		cin >> inputValue;
+	}
+	int currentValue;
+	int sizeOfStack = myStack.count();
```
Зачем эта переменная? Почему просто не используется myStack.count()?

eggorrus replied:
т.к. используется цикл for, то с каждой итерацией цикла он обращается к методу count стека, а у нас размер стека с каждой итерацией уменьшается, т.к. мы извлекаем значение. Можно это сделать так, как Вы предложили, тогда нужно делать i— каждый раз. Я посчитал нецелесообразным при каждой итерации обращаться к стеку

ichepkov2401 replied:
Тогда наверное было бы логичнее использовать while(myStack.count() > 0). вместо цикла с заданым количеством повторений

eggorrus replied:
сделан цикл while

eggorrus committed:
Исправлены замечания ревьюверов #4

1. Исправлена битая кодировка в комментариях
2. Для обработки стека теперь используется цикл while вместо for
3. В методе push более не используется ссылка

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,118 @@
+	/// Ниже представлен метод push. Он добавляет элемент на вершину стека.
+	/// Method push is below. It adds an element on top of the Stack.
+	/// </summary>
+	/// <param name="value"></param>
```
Пустые теги не нужны. Если в них ничего не пишите, удалите

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,118 @@
+	}
+	/// <summary>
+	/// Ниже представлен метод push. Он добавляет элемент на вершину стека.
+	/// Method push is below. It adds an element on top of the Stack.
```
Писать комментарии сразу на двух языках или что ещё хуже писать вперемешку не имеет особого смысла.  
Если вы подразумеваете, что с вашим кодом будут взаимодействовать только русскоязычные разработчик (например вы работаете в отечественной IT-компании и пишите не open source проект), то используйте русский язык. Если вы работаете над open source проектом или по крайней мере ваши контракты торчат наружу к иностранным разработчикам, пишите комментарии на английском.

eggorrus replied:
Исправлено

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,118 @@
+	void push (T value)
+	{
+		Node* newData = new Node(value);
+		newData->data = value;
```
Эта строчка не имеет смысла, вы уже положили данные туда через конструктор

ichepkov2401 commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,27 @@
+		cin >> inputValue;
+	}
+	cout << "Положительные числа в обратном порядке:" << endl;
+	while(myStack.count()>0)
```
```suggestion
	while(myStack.count() > 0)
```

ichepkov2401 commented on 2sem/lab1/lab1.cpp:
```diff
@@ -1,20 +1,27 @@
+	Stack<int> myStack;
+	cout << "Введите числа" << endl;
+	cin >> inputValue;
+	while (inputValue != 0)
```
Логический разделитель, чтобы код не выглядел полотном текста
```suggestion
	
	while (inputValue != 0)
```

eggorrus committed:
Исправлены замечания #5

1. Убраны комментарии на английском языке
2. Убраны пустые теги
3. Добавлены логические разделители, чтобы код не выглядел сплошным текстом

ichepkov2401 commented on 2sem/lab0/lab0.vcxproj:
```diff
@@ -129,6 +129,11 @@
   <ItemGroup>
     <ClCompile Include="lab0.cpp" />
   </ItemGroup>
+  <ItemGroup>
```
Этот проект не нужен

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,97 @@
+using namespace std;
+
+/// <summary>
+/// Ниже находится описание класса "Stack", который является системой хранения данных типа "LIFO"
```
Отсутствует закрывающий тег

ichepkov2401 replied:
Здесь и далее

ichepkov2401 commented on 2sem/common/stack.h:
```diff
@@ -0,0 +1,97 @@
+	/// <summary>
+	/// Конструктор для структуры Stack
+	Stack() : counter(0), top(nullptr) {}
+	/// <summary>
```
Разделители между функциональными членами класса
```suggestion

	/// <summary>
```

ichepkov2401 replied:
Здесь и далее

eggorrus committed:
Исправлены замечания #6

1. Удален проект lab0
2. Добавлены разделения функциональных членов класса
3. Прописаны закрывающие теги

ichepkov2401 commented on 2sem/lab0/lab0.vcxproj:
```diff
@@ -129,6 +129,11 @@
     <ClCompile Include="lab0.cpp" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\common\common.vcxproj">
```
Вы проект из решения исключили, а файлики с папкой остались. Их Надо удалить

eggorrus replied:
сделано

eggorrus committed:
Удалены файлы lab0
